### ISSUE TITLE
ImportUDF: Must hardcode python path to avoid Excel error '48' - file not found 'xlwings32.dll'

### ISSUE THREAD

**Issue by tschluck** [ZoomerAnalytics/xlwings#365]
I am running xlwings 0.6.4 on Excel 2010 (32-bit) and Canopy / Python 2.7. (32-bit).
When trying to import UDFs I always get the error '48' - file not found: xlwings32.dll
Only when I hardcode the Python path into the PYTHON_WIN variable the error is solved.
As I'd like to distribute my files later on I wonder if this issue can be somehow avoided.

Appreciate any input and help.

**Comment by fzumstein** [comment]
I need some more info:

* Did you install xlwings as Canopy package or via pip?
* Is Canopy your default Python installation, i.e. does it start by just typing `python` at the command prompt?

**Comment by tschluck** [comment]
Thanks for your answer. To your questions:
- Yes, Canopy is my standard Python installation. On the cmd it starts by just writing `python`.
- I installed xlwings via pip.
- `echo python` gives me two paths (see below) and in both folders I have the xlwings32.dll

> C:\Users\username\AppData\Local\Enthought\Canopy32\User
> C:\Users\username\AppData\Local\Enthought\Canopy32\User\Scripts

**Comment by fzumstein** [comment]
Get these paths also listed when your run `echo %path%`? the dlls are expected to sit in the same directory as the `python.exe`. Given that they occur in two directories, maybe there's some confusion between 2 installations? although I woulnd't have expected it to be an issue in terms of finding the dll.

**Comment by tschluck** [comment]
Yes, both path are listed in the path-variable.
I did reinstall Canopy once, because I had to downgrade it from 64-bit to 32-bit for compatibility issues. But before installing the 32-bit version I uninstalled the complete 64-bit version.
In both folders the python executable and the xlwings-dll are the same. Python 2.7.10. and identical in size. For the xlwings32.dll I can only tell by its size.
I will also write to Enthought and ask whether this double-path-installation is standard or if somethin went wrong.

**Comment by fzumstein** [comment]
It's difficult to tell what's going on, but Windows looks in the PATH env. variables for dlls, so if it is in one of the folders and this is listed on the PATH, then that *should* work.
Things you can check/do are: 

* Remove the dll that is *not* in the same folder as python.exe
* Make sure you use the latest version of the vba xlwings module
* Is the xlwings32.dll approx. 241KB (in previous versions, it was smaller)
* As usual when dlls/path variables are involved, a restart will sometimes do good

**Comment by tschluck** [comment]
Tried all of the above - no effect.
Looking closely through the path variable I found old entries from my 64-bit installation and removed them. But the more interesting thing is, that this first installation also had these two paths set. Therefore I don't think it is a  side effect of an installation gone wrong.
I've also seen, that I set the paths in the path variable of the user and not the systems. Hoping for some internal misbehavior I transferred the paths now to the systems variable, but also with no effect (and in both cases `echo %path%` gives the same result).

I also have seen that I run 32-bit Python, 32-bit Excel with VBA 7 on a 64-bit Windows 7 machine.
Could this be an issue? 
In the xlwings.bas it is checked for `VBA7` and `Win64` and when running the Debugger he throws me out in the `Py()`-function right after `XLPyLodaDLL`.
Marker at `If 0 <> XLPyDLLActivateAuto(Py, XLPyCommand) Then Err.Raise 1000, Description:=Py`

**Comment by fzumstein** [comment]
Since things work correctly when you point to the correct folder, I dare say that it's indeed a path issue. xlwings works with all bit combinations: depending on the bit version of Excel, it either looks for the xlwings32.dll or xlwings64.dll (python and windows bit version don't matter).

Some more ideas:
* Go through all the paths on `echo %path%` and make sure the xlwings32.dll is placed in only one of those paths (next to the default python.exe)
* Completely uninstall xlwings (Make sure there's no xlwings32.dll left anywhere) and reinstall with pip (or maybe try the Canopy package, I never tried that one).

Long story short: i think it's your system and so deploying is usually not an issue. I would also recommend looking into the Anaconda distribution (it already comes with xlwings included and sets the path correctly when installing with the defaults).

**Comment by tschluck** [comment]
Again, I tried all of your suggestions without any success. The complete uninstallation and pip reinstall I did yesterday. Now, I tried the same using the package manager from Enthought.
Well, I can live with it. 
BTW, when using the Enthought installer you have to copy the DLLs to one of the two python-directories manually. They land some below in the `...\User\Libs\site-packages\` - folder.
Nevertheless a big thanks for your support and I hope you keep going with xlwings.

**Comment by fzumstein** [comment]
Ok thanks for letting me know about the false Canopy package.
